winterize/init.lua

```diff
@@ -0,0 +1,102 @@
+local snowy_dirt_tiles = minetest.registered_nodes["default:dirt_with_snow"].tiles
+local snowy_dirt_sounds = minetest.registered_nodes["default:dirt_with_snow"].sounds
+local grasses = {"dry_grass", "grass", "coniferous_litter",}
+local leaves = {"leaves", "aspen_leaves", "jungleleaves"}
+
+minetest.register_node("winterize:ice", { -- breaks instantly, drops nothing
+	drawtype = "nodebox",
+	description = "Ice",
+	tiles = {"winterize_ice_seethrough.png"},
+	buildable_to = true,
+	floodable = true,
+	paramtype = "light",
+	sunlight_propogates = true,
+	node_box = {
+		type = "fixed",
+		fixed = {
+			{-0.5, -0.5, -0.5, 0.5, -0.4, 0.5},
+		},
+	},
+	collision_box = {
+		type = "fixed",
+		fixed = {
+			{-0.5, -0.5, -0.5, 0.5, -0.4, 0.5},
+		},
+	},
+	drop = "",
+	groups = {dig_immediate = 3, slippery = 4},
+	sounds = default.node_sound_glass_defaults(),
+})
+
+minetest.register_lbm({
+	label = "Add ice layer to top of water",
+	name = "winterize:top_water_with_ice",
+	nodenames = {"default:water_source"},
+	run_at_every_load = true,
+	action = function(pos, node)
+		local pos_above = vector.new(pos.x, pos.y + 1, pos.z)
+
+		if minetest.get_node(pos_above).name == "air" then
+			minetest.set_node(pos_above, {name = "winterize:ice"})
+		end
+	end
+})
+
+local function can_snow_fall_freely(pos)
+	local voxelmanip = VoxelManip()
+	local aircheck_pos = vector.new(pos.x, pos.y + 20, pos.z)
+	local vpos1, vpos2 = voxelmanip:read_from_map(pos, aircheck_pos)
+	local voxelarea = VoxelArea:new{MinEdge=vpos1, MaxEdge=vpos2}
+	local nodes_above = voxelmanip:get_data()
+
+	for ypos = pos.y, aircheck_pos.y, 1 do
+		local nodeid = nodes_above[voxelarea:indexp(vector.new(pos.x, ypos, pos.z))]
+		if nodeid ~= minetest.CONTENT_AIR and
+		minetest.registered_items[minetest.get_name_from_content_id(nodeid)].pointable then -- ignore barriers
+			return true -- Obstuction found
+		end
+	end
+
+	return false -- Snow can fall freely
+end
+
+minetest.register_lbm({
+	label = "Place snow on top of nodes",
+	name = "winterize:top_nodes_with_snow",
+	nodenames = {"group:crumbly", "group:leaves", "group:cracky", "group:choppy"},
+	run_at_every_load = true,
+	action = function(pos, node)
+		local pos_above = vector.new(pos.x, pos.y + 1, pos.z)
+
+		if minetest.get_node(pos_above).name == "air" and node.name ~= "default:snow" and
+		node.name ~= "ctf_map:cobble" and minetest.registered_items[node.name].walkable then
+			if not can_snow_fall_freely(pos_above) then
+				minetest.set_node(pos_above, {name = "default:snow"})
+			end
+		end
+	end
+})
+
+for _, leaftype in pairs(leaves) do
+	minetest.override_item("default:" .. leaftype, {
+		tiles = {"winterize_dead_leaves.png"},
+	})
+
+	minetest.override_item("ctf_map:" .. leaftype, {
+		tiles = {"winterize_dead_leaves.png"},
+	})
+end
+
+for _, grasstype in pairs(grasses) do
+	minetest.override_item("default:dirt_with_" .. grasstype, {
```

Comment:
I'm still not comfortable with overriding the actual nodes. I'd prefer replacing them with `default:dirt_with_snow` in an LBM instead.

Reply:
I think overriding is much easier here, more reliable and better performance wise. I also think that it's not a problem to make all those grass nodes snowy because usually where grass can grow, snow can also fall.

So, I actually prefer the current solution.

Reply:
Fine. I'm still neutral, but will approve once the function is renamed and the dependency fixed.